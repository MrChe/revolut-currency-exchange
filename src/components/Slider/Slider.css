.Slider {
    background-color: black !important;
}
