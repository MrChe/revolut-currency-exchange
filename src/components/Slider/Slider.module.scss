.Slider {
    position: relative;
    height: 100%;
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
}


:global {

    .swiper-wrapper {
        margin: 0;
        padding: 0;
        padding-inline-start: 0;
        list-style-type: none;
    }

    .swiper-container {
        width: 100%;
        height: 100%;
    }

    .swiper-slide {
        text-align: center;
        font-size: 18px;
        background-color: transparent;
        display: flex;
        justify-content: center;
        align-items: center;
        flex-direction: column;
    }

    .swiper-button-next, .swiper-button-prev {
        color: white;
    }

    .swiper-pagination-bullet-active {
        background-color: white;
    }

}

