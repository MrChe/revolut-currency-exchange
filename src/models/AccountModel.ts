import { action, makeAutoObservable, observable } from "mobx";
import { v4 as uuidv4 } from "uuid";
// import { persistStore } from "../utils/mobx-persist.utils";

export class AccountModel {
  public id: string;
  public currency: string;
  public balance: number;
  // public history: [];
  constructor(data: Partial<AccountModel>) {
    this.id = data.currency || uuidv4();
    this.currency = data.currency || "";
    this.balance = Number(data.balance?.toFixed(2)) || 100;
    // this.history = data.history;
    makeAutoObservable(this, {
      currency: observable,
      balance: observable,
      updateBalance: action,
    });
    // persistStore(this, ["id", "currency", "balance"], "AccountModel");
  }

  public updateBalance = (balance: number): void => {
    this.balance = Number(balance.toFixed(2));
  };
}
