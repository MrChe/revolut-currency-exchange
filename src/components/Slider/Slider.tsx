import React from "react";
import { Swiper, SwiperSlide } from "swiper/react";
import SwiperCore, {
  Navigation,
  Pagination,
  EffectCube,
  History,
} from "swiper";
import SwiperClass from "swiper/types/swiper-class";
import { AccountModel } from "../../models/AccountModel";
import { Account } from "../../components/Account/Account";
import { HistoryOptions } from "swiper/types/components/history";
import { formatCurrency, currencyNames } from "../../utils/helpers";

import "swiper/swiper-bundle.min.css";
import "swiper/components/navigation/navigation.min.css";
import "swiper/components/pagination/pagination.min.css";
import "swiper/components/scrollbar/scrollbar.min.css";

import styles from "./Slider.module.scss";

// install Swiper modules
SwiperCore.use([Navigation, Pagination, EffectCube, History]);

interface ISliderProps {
  data: AccountModel[];
  id: string;
  history?: HistoryOptions | boolean;
  onSlideChange?: (swiper: SwiperClass) => void;
  onSwiper?: (swiper: SwiperClass) => void;
  view: "preview" | "exchange";
}

export const Slider = (props: ISliderProps): JSX.Element => {
  return (
    <div className={styles.Slider}>
      <Swiper
        history={props.history}
        id={props.id}
        tag={"section"}
        wrapperTag={"ul"}
        centeredSlides
        navigation={false}
        pagination={{
          clickable: true,
        }}
        spaceBetween={0}
        slidesPerView={1}
        onSwiper={props.onSwiper}
        onSlideChange={props.onSlideChange}
      >
        {props.data.map((account: AccountModel) => {
          return (
            <SwiperSlide tag={"li"} key={account.id} data-history={account.id}>
              <Account
                balance={formatCurrency(account.balance, account.currency)}
                view={props.view}
                currency={account.currency}
                currencyName={currencyNames[account.currency] || ""}
              />
            </SwiperSlide>
          );
        })}
      </Swiper>
    </div>
  );
};
