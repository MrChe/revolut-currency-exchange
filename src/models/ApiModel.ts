import Axios, { AxiosInstance } from "axios";
import { IRates } from "../models/ExchangeModel";
import { filterRates } from "../utils/currencies";

const APP_ID = "e285ab8ceb004923b3e423a0d0583915";

export class ApiModel {
  private client: AxiosInstance;
  constructor() {
    this.client = Axios.create({
      baseURL: `https://openexchangerates.org/api`,
      params: {
        base: "USD",
        app_id: APP_ID,
        symbols: "USD,EUR,GBP,UAH",
      },
      headers: {
        "Content-Type": "application/json",
      },
    });
  }

  public getLatestRatesRequest = async (): Promise<IRates | undefined> => {
    try {
      const response = await this.client.get(`/latest.json`);
      if (!response.data) {
        throw new Error("Not have any data");
      }
      return response.data;
    } catch (error) {
      console.error("getLatestRatesRequest", error);
    }
  };
}
