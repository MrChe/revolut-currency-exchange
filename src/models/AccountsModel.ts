import { RootModel } from "./RootModel";
// import { persistStore } from "../utils/mobx-persist.utils";
import { action, computed, makeAutoObservable, observable } from "mobx";
import { AccountModel } from "./AccountModel";
import { IRates } from "./ExchangeModel";

export class AccountsModel {
  rootModel: RootModel;
  public accounts: AccountModel[];
  public ratesData: IRates | null;
  public selectedAccount: AccountModel | null;
  constructor(rootModel: RootModel, currencies: string[], base: string) {
    this.rootModel = rootModel;
    this.selectedAccount = null;
    this.accounts = [];
    this.ratesData = null;
    makeAutoObservable(this, {
      // rootModel: false,
      accounts: observable,
      selectedAccount: observable,
      accountsAsArray: computed,
      setSelectedAccount: action,
    });

    // INIT Module
    this.init(currencies, base);
    // persistStore(
    //   this,
    //   ["accounts", "selectedAccount", "accountsAsArray"],
    //   "AccountsModel",
    // );
  }

  private init = async (currencies: string[], base: string): Promise<void> => {
    try {
      if (currencies.length !== 0) {
        const data = await this.rootModel.ApiModel.getLatestRatesRequest(
          currencies,
          base,
        );
        if (data) {
          this.setRatesData(data);
          this.setAccounts(data);
        }
      } else {
        throw new Error("Please add some currencies like ['USD', 'UAH']");
      }
    } catch (error) {
      console.error("init", error);
    }
  };

  public setSelectedAccount = (id: string): void => {
    const foundAccount = this.getAccountById(id);
    if (foundAccount) {
      this.selectedAccount = foundAccount;
    } else {
      console.error(`Cant find account by id: ${id}`);
    }
  };

  public get accountsAsArray(): AccountModel[] {
    return this.accounts.slice();
  }

  public setRatesData = (ratesData: IRates): void => {
    this.ratesData = ratesData;
  };

  public setAccounts = (ratesData: IRates): void => {
    Object.keys(ratesData.rates).forEach((currency: string) => {
      this.accounts.push(
        new AccountModel({
          currency,
          balance: 100,
        }),
      );
    });
  };

  public getAccountById = (id: string): AccountModel | undefined => {
    return this.accounts.find((account) => account.id === id);
  };
}
