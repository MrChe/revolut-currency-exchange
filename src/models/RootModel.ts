import { ExchangeModel } from "./ExchangeModel";
import { AccountsModel } from "./AccountsModel";
import { ApiModel } from "./ApiModel";

export class RootModel {
  public ExchangeModel: ExchangeModel;
  public AccountsModel: AccountsModel;
  public ApiModel: ApiModel;
  constructor() {
    this.ApiModel = new ApiModel();
    // init accounts
    this.AccountsModel = new AccountsModel(
      this,
      ["USD", "EUR", "GBP", "UAH"],
      "USD",
    );
    this.ExchangeModel = new ExchangeModel(this);
  }
}

export const rootModel = new RootModel();
