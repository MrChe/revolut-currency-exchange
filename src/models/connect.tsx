// holds a reference to the store (singleton)
import { createContext, useContext } from "react";
import { RootModel } from "./RootModel";

// create the context
const MobxContext = createContext<RootModel | null>(null);

// create the provider component
export const MobxProvider = MobxContext.Provider;

// create the hook
export const useStore = (): RootModel => {
  const context = useContext(MobxContext);
  if (!context) {
    throw new Error("useStore must be used within MobxContext");
  }

  return context;
};
