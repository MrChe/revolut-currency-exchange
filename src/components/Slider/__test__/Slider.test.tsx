import React from "react";
import renderer from "react-test-renderer";
import { Slider } from "../Slider";
import { TestWrapper } from "../../../../tests/testWrapper";
import { rootModel } from "../../../models/RootModel";
import { ratesData } from "../../../../tests/mockData";

describe("Slider Component", () => {
  const ExchangeModel = rootModel.ExchangeModel;
  beforeAll(() => {
    ExchangeModel.setRatesData(ratesData);
    ExchangeModel.setAccounts(ratesData);
    ExchangeModel.setActiveAccountFrom(ExchangeModel.accounts[0].id);
    ExchangeModel.setActiveAccountTo(ExchangeModel.accounts[1].id);
  });
  it("Slider for dashboard snapshot", () => {
    const handleSlideChange = jest.fn();
    const onInitSwiper = jest.fn();
    const tree = renderer
      .create(
        <TestWrapper>
          <Slider
            history={{
              key: "dashboard",
            }}
            data={ExchangeModel.accountsAsArray}
            id={"dashboard_slider"}
            onSlideChange={handleSlideChange}
            onSwiper={onInitSwiper}
            view={"preview"}
          />
        </TestWrapper>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
