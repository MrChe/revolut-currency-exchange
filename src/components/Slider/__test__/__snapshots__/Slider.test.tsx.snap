// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider Component Slider for dashboard snapshot 1`] = `
<div
  className="Slider"
>
  <section
    className="swiper-container"
    id="dashboard_slider"
  >
    <div
      className="swiper-pagination"
    />
    <ul
      className="swiper-wrapper"
    >
      <li
        className="swiper-slide"
        data-history={0}
      >
        <div
          className="Account View_Preview"
        >
          <div
            className="CurrencyInformation"
          >
            <div
              className="Wrapper"
            >
              <div>
                <p
                  className="CurrencyTitle"
                >
                  €100.00
                </p>
                <p>
                  Euro
                </p>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        className="swiper-slide"
        data-history={1}
      >
        <div
          className="Account View_Preview"
        >
          <div
            className="CurrencyInformation"
          >
            <div
              className="Wrapper"
            >
              <div>
                <p
                  className="CurrencyTitle"
                >
                  £100.00
                </p>
                <p>
                  British Pound Sterling
                </p>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        className="swiper-slide"
        data-history={2}
      >
        <div
          className="Account View_Preview"
        >
          <div
            className="CurrencyInformation"
          >
            <div
              className="Wrapper"
            >
              <div>
                <p
                  className="CurrencyTitle"
                >
                  ₴100.00
                </p>
                <p>
                  Ukrainian Hryvnia
                </p>
              </div>
            </div>
          </div>
        </div>
      </li>
      <li
        className="swiper-slide"
        data-history={3}
      >
        <div
          className="Account View_Preview"
        >
          <div
            className="CurrencyInformation"
          >
            <div
              className="Wrapper"
            >
              <div>
                <p
                  className="CurrencyTitle"
                >
                  $100.00
                </p>
                <p>
                  United States Dollar
                </p>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </section>
</div>
`;
